### DISCUSSION TITLE
Filter parents by matching multiple child fields within same child record

### DISCUSSION THREAD

**Discussion by hashan-aet** [directus/directus#13828] [Archive]
I have a data structure like this in Directus,

```json
[
   {
      "category":"Phone",
      "id":"29121a24-2315-40f1-bee5-7f09f2279a8e",
      "name":"iPhone 13",
      "variants":[
         {
            "name":"256 GB",
            "stocks":[
               {
                  "qty":2,
                  "store":"Store 1"
               },
               {
                  "qty":5,
                  "store":"Store 2"
               }
            ]
         }
      ]
   },
   {
      "category":"Laptop",
      "id":"ee0e9705-d431-4bb2-ae09-92cbad237f7c",
      "name":"Macbook Air 2020",
      "variants":[
         {
            "name":"13 inch",
            "stocks":[
               {
                  "qty":3,
                  "store":"Store 2"
               },
               {
                  "qty":0,
                  "store":"Store 1"
               }
            ]
         }
      ]
   }
]
```

In here I just want to get all products for the **Store 1** if the quantity is greater than zero, so I tried below filters,

**Filter 1**: **{"variants": {"stocks": {"store": "Store 1", "qty": {"_gt": 0}}}}**
This filter returns both products, condition on qty field did not applied.

![Postman 6_10_2022 2_38_20 PM](https://user-images.githubusercontent.com/103099031/173032599-5b4330ca-c0fb-49b3-a89c-fff850074493.png)

**Filter 2**: **{"_and": [{"variants":{"stocks": {"store": "Store 1"}}}, {"variants":{"stocks": {"qty": {"_gt": 0}}}}]}**
This also returns both, here the first condition has checked on the first stock record and the second condition has checked on the second stock record. Both conditions did not check within the same record.

![Postman 6_10_2022 2_39_23 PM](https://user-images.githubusercontent.com/103099031/173032542-78504bc1-8281-4458-adfa-0b5a996e4057.png)

Is there any other way to write this filter to get the expected output?

**Comment by rijkvanzanten** [comment]
Hmm good one. The second one should be correct. Could you share a database dump / schema snapshot so we can reproduce to test with?

**Reply by hashan-aet** [reply]
Please find database dump [here](https://we.tl/t-TxcV586RPS).

**URL with Filter 1:** `http://localhost:8055/items/my_product?fields=*.*,variants.stocks.*&filter={"variants": {"stocks": {"store": "Store 1", "qty": {"_gt": 0}}}}`

**URL with Filter 2:** `http://localhost:8055/items/my_product?fields=*.*,variants.stocks.*&filter={"_and": [{"variants":{"stocks": {"store": "Store 1"}}}, {"variants":{"stocks": {"qty": {"_gt": 0}}}}]}`

**Note:** I experienced this issue only when applying the filter for parameters in a child of a child(product -> variant -> **stock**)

**Reply by rijkvanzanten** [reply]
Dump reuploaded for archival reasons: [directus-dump.sql.zip](https://github.com/directus/directus/files/8891715/directus-dump.sql.zip)

**Reply by rijkvanzanten** [reply]
This is currently working as expected, though doesn't get you the result you're looking for. The _and on the top level means it will use a SQL in query similar to

```sql
SELECT x
FROM y
WHERE 
  id IN ((/* select ids for filter 1 */) AND (/* select ids for filter 2 */))
```

whereas your result would require

```sql
SELECT x
FROM y
WHERE 
  id IN (/* select ids for filter 1 AND filter 2 */)
```

as they relate to the same join path. This is the difference between your example 1 and example 2, though only example 2 is currently supported.

As for right now, I would recommend starting from `my_product_variant_stock`, as the query _I just want to get all products for the Store 1 if the quantity is greater than zero_ can be solved with:

```
GET /items/my_product_variant_stock
  ?fields=*,variant_id.product_id
  &filter[qty][_gte]=1
  &filter[store][_eq]=Store 1
```
